test/e2e/storage/regional_pd.go

```diff
@@ -0,0 +1,346 @@
+		framework.Logf("deleting claim %q/%q", pvc.Namespace, pvc.Name)
+		framework.ExpectNoError(c.CoreV1().PersistentVolumeClaims(pvc.Namespace).Delete(pvc.Name, nil),
+			"Error deleting claim %s.", pvc.Name)
+		err = framework.WaitForPersistentVolumeDeleted(c, pvc.Spec.VolumeName, framework.Poll, pvDeletionTimeout)
+		if err != nil {
+			framework.Logf("WARNING: PV %s is not yet deleted, and subsequent tests may be affected.", pvc.Spec.VolumeName)
+		}
+	}()
+
+	By("getting zone information from pod")
+	pod := getPod(c, ns, repdLabels)
+	nodeName := pod.Spec.NodeName
+	node, err := c.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
+	Expect(err).ToNot(HaveOccurred())
+	podZone := node.Labels[apis.LabelZoneFailureDomain]
+
+	// TODO (verult) Consider using node taints to simulate zonal failure instead.
+	By("deleting instance group belonging to pod's zone")
+	cloud, err := framework.GetGCECloud()
+	if err != nil {
+		Expect(err).NotTo(HaveOccurred())
+	}
+	instanceGroupName := framework.TestContext.CloudConfig.NodeInstanceGroup
+	instanceGroup, err := cloud.GetInstanceGroup(instanceGroupName, podZone)
+	Expect(err).NotTo(HaveOccurred(),
+		"Error getting instance group %s in zone %s", instanceGroupName, podZone)
+	err = framework.DeleteManagedInstanceGroup(podZone)
+	Expect(err).NotTo(HaveOccurred(),
+		"Error deleting instance group in zone %s", podZone)
+	defer func() {
+		framework.Logf("recreating instance group %s", instanceGroup.Name)
+
+		// HACK improve this when Managed Instance Groups are available through the cloud provider API
+		templateName := strings.Replace(instanceGroupName, "group", "template", 1 /* n */)
+
+		framework.ExpectNoError(framework.CreateManagedInstanceGroup(instanceGroup.Size, podZone, templateName),
+			"Error recreating instance group %s in zone %s", instanceGroup.Name, podZone)
+		framework.ExpectNoError(framework.WaitForReadyNodes(c, nodeCount, framework.RestartNodeReadyAgainTimeout),
+			"Error waiting for nodes from the new instance group to become ready.")
+	}()
+
+	// Rescheduling of the pod should occur while instance group deletion is in progress.
+	// If this is no longer the case, need to verify the pod deleted before waiting for the
+	// statefulset to become ready.
+	framework.WaitForStatefulSetReplicasReady(statefulSet.Name, ns, c, 3*time.Second, framework.RestartPodReadyAgainTimeout)
+
+	By("verifying the same PVC is used by the new pod")
+	Expect(getPVC(c, ns, repdLabels).Name).To(Equal(pvc.Name),
+		"The same PVC should be used after failover.")
+
+	By("verifying the container output has 2 lines, indicating the pod has been created twice using the same RePD.")
+	pod = getPod(c, ns, repdLabels)
+	logs, err := framework.GetPodLogs(c, ns, pod.Name, "")
+	Expect(err).NotTo(HaveOccurred(),
+		"Error getting logs from pod %s in namespace %s", pod.Name, ns)
+	lineCount := len(strings.Split(strings.TrimSpace(logs), "\n"))
+	expectedLineCount := 2
+	Expect(lineCount).To(Equal(expectedLineCount),
```

Comment:
Would it be too fancy for the Pod to print out the zone it's in, and validate that?

Reply:
Is there a way for a pod to get its zone information other than to `kubectl get node` and check its label? We are already kind of doing that in the test.

Reply:
I'm not sure if downward api can expose labels.

Reply:
Had a chat with @msau42 offline - pods usually don't expose zone labels by default, so it's a bit more involved to get zone info within a container.